src/workshed/utils.ts

```diff
@@ -0,0 +1,178 @@
+}
+
+function getTrainsetPositionsUntilConfigurable(): number {
+  const pos = toInt(get("trainsetPosition", ""));
+  const configured = toInt(get("lastTrainsetConfiguration", ""));
+  const turnsSinceConfigured = pos - configured;
+
+  return Math.max(0, 40 - turnsSinceConfigured);
+}
+
+export function isTrainsetConfigurable(): boolean {
+  return getWorkshed() === $item`model train set` && getTrainsetPositionsUntilConfigurable() <= 0;
+}
+
+export function setTrainsetConfiguration(pieces: TrainsetPiece[]): void {
+  visitUrl("campground.php?action=workshed");
+
+  const pieceIds = pieces.map((piece, index) => `slot[${index}]=${getPieceId(piece)}`);
+
+  const url = `choice.php?forceoption=0&whichchoice=1485&option=1&pwd&${pieceIds.join("&")}`;
+
+  visitUrl(url, true);
+  visitUrl("main.php");
+
+  const expected = pieces.join(",");
+
+  if (expected !== get("trainsetConfiguration", "")) {
+    throw new Error(
+      `Expected trainset configuration to have changed, expected "${expected}" but instead got ${get(
+        "trainsetConfiguration",
+        ""
+      )}`
+    );
+  }
+}
+
+export const defaultPieces = [
+  TrainsetPiece.DOUBLE_NEXT_STATION,
+  TrainsetPiece.GAIN_MEAT,
+  TrainsetPiece.CANDY,
+  TrainsetPiece.RANDOM_BOOZE,
+  TrainsetPiece.DROP_LAST_FOOD_OR_RANDOM,
+  TrainsetPiece.ORE,
+  TrainsetPiece.EFFECT_MP,
+  TrainsetPiece.GAIN_STATS,
+];
+
+export function grabMedicine(): void {
+  const options = visitUrl("campground.php?action=workshed");
+  let i = 0;
+  let match;
+  const regexp = /descitem\((\d+)\)/g;
+  const itemChoices = new Map<Item, number>();
+  if (!globalOptions.nobarf) {
+    // if spending turns at barf, we probably will be able to get an extro so always consider it
+    itemChoices.set($item`Extrovermectin™`, -1);
+  }
+
+  while ((match = regexp.exec(options)) !== null) {
+    i++;
+    const item = descToItem(match[1]);
+    itemChoices.set(item, i);
+  }
+
+  const bestItem = maxBy([...itemChoices.keys()], garboValue);
+  const bestChoice = itemChoices.get(bestItem);
+  if (bestChoice && bestChoice > 0) {
+    visitUrl("campground.php?action=workshed");
+    runChoice(bestChoice);
+  }
+  if (handlingChoice()) visitUrl("main.php");
+}
+
+export function getTrainsetConfiguration(): TrainsetPiece[] {
+  return get("trainsetConfiguration").split(",") as TrainsetPiece[];
+}
+
+export function offsetDefaultPieces(offset: number): TrainsetPiece[] {
+  const newPieces: TrainsetPiece[] = [];
+  for (let i = 0; i < 8; i++) {
+    const newPos = (i + offset) % 8;
+    newPieces[newPos] = defaultPieces[i];
+  }
+  return newPieces;
+}
+
+export function stringToWorkshedItem(): Item {
+  // An empty string is a subset of every string and will match all the worksheds
+  // So we explicitly handle this case here
+  if (globalOptions.workshed === "") return $item`none`;
+
+  const lowerCaseWorkshed = globalOptions.workshed.toLowerCase();
+  const validWorksheds = [
+    ...new Map([
```

Comment:
I don't think there's any utility in wrapping this chunk in the map constructor and then calling entries, we never use the mappiness of it.